### ISSUE TITLE
Add a Pulumi provider equivalent to the Kestra Terraform provider

### ISSUE THREAD

**Issue by anna-geller** [kestra-io/kestra#3016]
### Feature description

Request from a Slack community user: "I think it would be interesting to port the Terraform Kestra plugin to Pulumi,
it can be done easily via this tool: https://github.com/pulumi/pulumi-terraform-bridge"

**Comment by anna-geller** [comment]
We may be able to use https://github.com/pulumi/pulumi-tf-provider-boilerplate as a getting-started point

Requested a.o. by @japerry911

**Comment by anna-geller** [comment]
@Ben8t this plugin has most upvotes among all plugin issues, worth prioritizing

**Comment by Ben8t** [comment]
I tested running the setup explained in https://github.com/pulumi/pulumi-tf-provider-boilerplate:

I'm facing some import errors regarding github.com/kestra-io/terraform-provider-kestra

1. From the error message it seems that terraform-provider-kestra is a program (`package main`) and can't be imported as module 🤔 

![Image](https://github.com/user-attachments/assets/00b6e3c4-2960-4e6c-afca-5461045f62e1)

2. When trying to import underlying sources (`internal/provider`) I hit classic private/public import error:

![Image](https://github.com/user-attachments/assets/3a0a1584-7277-4447-b108-35699767c75f)

cc @Skraye @tchiotludo if you have any idea I can look further

**Comment by tchiotludo** [comment]
@Ben8t I move to unknown version, we don't have any others users asking for pulumi, the only one we know moved to terraform and is happy with

**Comment by anna-geller** [comment]
This will need to wait until we have a Go SDK

**Comment by antoine-mle** [comment]
Hi, I would be happy to have an equivalent plugin with OpenTofu or Pulimi 😉

**Comment by munjalpatel** [comment]
```
pulumi package add terraform-provider kestra-io/kestra
```

^^ works quite well using [Pulumi Any Terraform Provider](https://www.pulumi.com/blog/any-terraform-provider)

**Comment by fdelbrayelle** [comment]
Regarding the last comment with the working command, does it still worth it @Ben8t @anna-geller @tchiotludo or should we close the issue? I feel like it does the job without creating a dedicated Pulumi package here.

**Comment by anna-geller** [comment]
yes, but before closing, this would need to be validated by someone internally; @fdelbrayelle  feel free to take a first stab at https://github.com/kestra-io/docs/issues/2659 and let Will + AJ continue from your starting point 👍

**Comment by japerry911** [comment]
@anna-geller , we don't use Pulumi, but I see this is a high ranked/needed issue, and would be helpful for the community. I feel confident I can complete this; as a result, I can make an attempt at this this upcoming Saturday. 

If that works, could you please assign to me?

**Comment by fdelbrayelle** [comment]
Hi Jack 👋 Thank you for your proposal! I'll assign it to you then 🚀

**Comment by japerry911** [comment]
EDITED - 

Just kidding, I read all of the comments and pieced it today.. Validate/How-To

**Comment by japerry911** [comment]
I was able to get Namespace to create, but Flows do not. They fail with 401, unauthorized..

```
 error: status: 401, method: POST, body: {"message":"Unauthorized","logref":null,"path":null,"_links":{"self":{"href":"/api/v1/pulumi/flows","templated":false,"profile":null,"deprecation":null,"title":null,"hreflang":null,"type":null,"name":null}},"_embedded":{}}: 
```

Using 1.0.1.

Here is my Pulumi code, 
```python
"""A Python Pulumi program"""

import pulumi
import pulumi_kestra as kestra

# Provider
kestra_provider = kestra.Provider(
    "kestra_provider",
    api_token="<REDACTED>",
    url="<REDACTED>",
    tenant_id="pulumi",
    timeout=30,
)

# Namespace
hello_world_namespace = kestra.Namespace(
    resource_name="hello_world_namespace",
    namespace_id="hello_world_namespace",
    description="An example namespace.",
    opts=pulumi.ResourceOptions(provider=kestra_provider, depends_on=kestra_provider),
)

# Flow
kestra.Flow(
    resource_name="hello_world_flow",
    namespace=hello_world_namespace.namespace_id,
    flow_id="hello_world_flow",
    content="""
inputs:
  - id: your_name
    displayName: Your Name
    type: STRING
    required: true

tasks:
  - id: greet_name
    type: io.kestra.plugin.core.log.Log
    message: "Hello {{ inputs.your_name }}! 🚀"
    """,
    opts=pulumi.ResourceOptions(provider=kestra_provider, depends_on=kestra_provider),
)
```

My token II am using was admin-level too.

@munjalpatel did you have luck with creating Flows?

**Comment by AcevedoR** [comment]
FIY you might have 401 errors when creating flows if the request headers/content-type/body is considered invalid

**Comment by japerry911** [comment]
> FIY you might have 401 errors when creating flows if the request headers/content-type/body is considered invalid

Thank you @AcevedoR , I'll see if I can investigate it further when I am free again 🙏

**Comment by japerry911** [comment]
@AcevedoR  thank you sir, adding ID and namespace to top fixed it.. the docs [here](https://registry.terraform.io/providers/kestra-io/kestra/latest/docs/resources/flow) show content without it, which seems to be inaccurate. 

Thanks again, ill keep progressing on this issue :D

**Comment by anna-geller** [comment]
thank you Jack, fixed https://github.com/kestra-io/terraform-provider-kestra/commit/29a1580608345771da2c6d60723fc627ac2580c3 🙏

**Comment by anna-geller** [comment]
@japerry911 did you manage to make some progress here? if not, no worries, we could also unassign and let someone else have a look here

**Comment by japerry911** [comment]
@anna-geller yes I did! sorry went radio silent, got super busy :D 

If I draft up a blogpost this weekend on medium showcasing how to use [Pulumi Any Terraform Provider](https://www.pulumi.com/blog/any-terraform-provider) with Kestra Terraform Provider, will that fulfil this issue, or need anything else?

**Comment by anna-geller** [comment]
yes absolutely! ☺️ thanks so much in advance! 🙏

**Comment by anna-geller** [comment]
draft is here so I'll close already https://jackskylord.medium.com/1a7a62c8c380

As discussed via DM, let's publish the blog and optionally if you'd like let's copy the same to our docs repo, either to How-to Guides or e.g. to this page https://kestra.io/docs/version-control-cicd. Much appreciated 🙏
